### ISSUE TITLE
Declare previous version conflict with dep

### ISSUE THREAD

**Issue by XedinUnknown** [composer/composer#7194]
Hi,

This seems to me like a problem that should come up a lot, but I've not managed to find documentation for it. I have identified that a previous version of my package conflicts with another dependency at a specific version. In my specific case, a previous release doesn't work with PHP 5.4.0 due to a bug in PHP related to trait composition and abstract methods. How can I now declare that not my current version (upcoming release), but my `0.1-alpha2` conflicts with PHP 5.4.0?

I understand that this may be difficult to do via Composer config. But maybe this could be done on Packagist?

#### Note
According to [SemVer 2.0 point 3][semver-20-3],

> Once a versioned package has been released, the contents of that version MUST NOT be modified. Any modifications MUST be released as a new version.

This means that re-tagging, i.e. making a modification and moving the tag to that commit, is a breach of SemVer.

[semver-20-3]: https://semver.org/spec/v2.0.0.html#spec-item-3

**Comment by alcohol** [comment]
No.

**Comment by AnrDaemon** [comment]
This is certainly can be solved… in your package. Declare that it requires 'php' above certain version and re-tag.

Not a composer issue.

**Comment by XedinUnknown** [comment]
Re-tag? That sounds like a very bad idea. If other packages depend on a state of your package, that state should never change. Is this really the only way?

@alcohol, what can you say about the proposed solution? Or could you suggest a solution of your own? Does this sound like a valid problem to you? Do you think this may be a feature in the future?

**Comment by alcohol** [comment]
> Do you think this may be a feature in the future?

I doubt it. Maybe in a future new major release. But that is a long way from happening. It would require changes in both Composer and Packagist, for what honestly is, a very rare edge-case scenario. Hardly no value vs risk/investment involved.

**Comment by AnrDaemon** [comment]
Yes, re-tag.
No, your package behavior does not change. Only its metadata (`composer.json`) do.
I don't see it as a problem.

**Comment by XedinUnknown** [comment]
Thanks, gentlemen.

**Comment by alcohol** [comment]
Retagging is not ideal. But probably the recommend approach. You might have to manually trigger an update in Packagist though I think. Not sure, I can't recall the last time I re-tagged something.

**Comment by XedinUnknown** [comment]
Just wanted to remind you that according to [SemVer 2.0 point 3][semver-20-3],

> Once a versioned package has been released, the contents of that version MUST NOT be modified. Any modifications MUST be released as a new version.

[semver-20-3]: https://semver.org/spec/v2.0.0.html#spec-item-3

**Comment by AnrDaemon** [comment]
The package *contents* are not changed.
`composer.json`, strictly speaking, is not "contents" of a package. (Many packages don't even HAVE it in dist archive, to begin with.)

**Comment by XedinUnknown** [comment]
Many packages don't have a `ContainerInterface.php` in them. But for those that do, it is part of the contents of the package, just like any documentation, or other any other files. Nowhere does it say that only PHP code is considered to be "contents" of a Composer/Packagist package. IMHO, the principle described by SemVer is there to ensure immutability of any particular release: if the commit is different, then anything can be different; if `composer.json` is different, Dog knows what else has changed. Also, according to [RFC 2119][], which I believe is used in the SemVer 2.0 spec, the word "MUST" means that "the definition is an absolute requirement of the specification", i.e. it is not optional under any circumstances.

[RFC 2119]: https://www.ietf.org/rfc/rfc2119.txt

**Comment by AnrDaemon** [comment]
Understand already, that release (code) errors and packaging errors are not the same.

**Comment by XedinUnknown** [comment]
That's true. But a package is a package, whether we talk about code in the package, or some meta or docs. If we update `package.json`, next someone will say "I only updated the inline doc", "I only added a dependency", "I only added an optional method argument" etc. But the release MUST be immutable, according to SemVer. Now, either SemVer is broken, or Composer does not follow 2.0 specs. And the latter I have nothing against, as long as the problem is acknowledged - which is: if the contents of a release can change, then the release is not dependable, i.e. there is no guarantee that the same consuming code will work with the same exact version of a dependency at two different points in time. But there definitely should be such a guarantee IMHO.

**Comment by alcohol** [comment]
There is a lot of room for interpretation here. Shall we just leave it at that?

**Comment by XedinUnknown** [comment]
@alcohol, not trying to argue; just want to get the problem acknowledged. Of course, we can just leave it at that, but it would be deeply disturbing.

I'd say, the topic is still open, and more input would be very useful. Perhaps, as a result of this discussion we will understand that the problem is there, but it should be solved in a different way: by updating SemVer, or Packagist, or changing some other definition somewhere, or updating Composer docs etc. It is my personal opinion that the problem should be solved with Composer, but I don't pretend that it is the only way, or the right way. I know I had closed the issue, but I would love to see it re-opened, and for the discussion to continue.

**Comment by AnrDaemon** [comment]
Your error in packaging your release should be resolved in Composer? I understand it right?

**Comment by XedinUnknown** [comment]
I found a bug in a release weeks after that release was made. This can be interpreted as a packaging error, I guess. But to me, given that the Composer config is used to indicate conflicts, it doesn't look that way. What do you think this is a concern of, @AnrDaemon?

**Comment by alcohol** [comment]
Since we don't require signing of releases, and all code is hosted on third party services and managed by third party developers, we can never guarantee that versions are immutable.

**Comment by XedinUnknown** [comment]
However, the commit hash does kinda guarantee that. If the commit, to which the version points, is the same at any time, then the state the same (save for pigeonhole problems, which is a different concern). If the commit is different at a different point in time, then the state is likely to not be the same.

**Comment by alcohol** [comment]
We deal with versions though, not commits.

**Comment by XedinUnknown** [comment]
And a version MUST be immutable. Pointing to a different commit breaks that.

**Comment by AnrDaemon** [comment]
You made up this claim.

**Comment by XedinUnknown** [comment]
Which? that a version is immutable? It's in the SemVer spec.

**Comment by AnrDaemon** [comment]
There's no mention of commits in semver specs.

**Comment by alcohol** [comment]
> And a version MUST be immutable. Pointing to a different commit breaks that.

Sure, but again, we do not deal with commits, we deal with versions. Commits are out of our scope/control.

**Comment by XedinUnknown** [comment]
I guess, we've all said what we had to. Now it would be great to hear other people's opinions. Would you mind if the issue is re-opened, maybe with a `feature-request` label or whatever you think is appropriate?

**Comment by alcohol** [comment]
I think I already answered that question though. It isn't going to happen. You'll need to simply tag a new version which includes the conflict rule and wait for people to adopt this new version. It would help if you have a place where you can broadcast news about this new version. For example a blog, twitter, or whatever.

**Comment by XedinUnknown** [comment]
I thought you said that there's a chance a feature like this will make it into the next major release. Would be great to know that there can be a better approach than re-tagging.

**Comment by alcohol** [comment]
Let me rephrase then, it **very likely** isn't going to happen.

**Comment by XedinUnknown** [comment]
Great. But maybe with more input, something will change. Whether it will happen or not, there is a need for something like this in general, and therefore I think this should remain open as a feature request. I'm sure some solution will be found somewhere.

**Comment by Jean85** [comment]
I referenced this issue in a PR, because we had this issue in liip/LiipFunctionalTestBundle: it was tagged as compatible with Symfony 4 but it wasn't, and we decided to allow it only in the newer major version.

Maybe Composer could allow declaring a single version as abandoned or deprecated? So during install it could warn the user, without deleting/changing the release (and breaking SemVer) and without a big BC break in Composer/Packagist.

Yes, it would be a big feature, but IMHO it could be doable without waiting for a major release here... What do you think?

**Comment by AnrDaemon** [comment]
> Maybe Composer

When you all stop confusing packagist with composer?

**Comment by Jean85** [comment]
Sorry, my bad.

**Comment by alcohol** [comment]
I am not sure how desirable it is to allow Packagist to provide additional metadata other than what can be found at the source (repositories).

@Seldaek do you have any thoughts on this?

**Comment by Seldaek** [comment]
If you published a bad version, just publish a fix version (n+1, avoid reusing the same tag) and move on. In the example above of the LiipFunctionalTestBundle it's a bit problematic because the old version might get picked if you use Symfony 4 at least until you release a version supporting Symfony 4. In that case maybe deleting the tag from the git repo and deleting the tag on packagist is the best course of action. People that have it in their composer.lock will still be able to install as the commit is still there and the lock file has the commit reference, but you stop the bad version from spreading further at least.

I don't really see what we can do on our end to mitigate this in a meaningful way that wouldn't take more effort than it's worth. Such mistakes don't happen all that often.
